Initial support for decoding Toshiba A/C remote messages.

crankyoldgit opened the pull request:
Ref #339

FYI @mwildbolz for additional review & testing.

crankyoldgit committed:
Initial support for decoding Toshiba A/C remote messages.
For #339

mwildbolz commented:
@crankyoldgit I implemented this with a minor change (RPT Mark timing) and was able to decode a real remote signal. But there is an issue with repeated decoding.
If i send a second signal with my remote, i get a memory alloc problem on my ESP causing it to restart, it just works once?!?

```
check_poison_block is called for free block 0x3fff12b8
Fatal exception 3(LoadStoreErrorCause):
epc1=0x4010011d, epc2=0x00000000, epc3=0x00000000, excvaddr=0x40031af8, depc=0x00000000

Exception (3):
epc1=0x4010011d epc2=0x00000000 epc3=0x00000000 excvaddr=0x40031af8 depc=0x00000000

ctx: cont 
sp: 3fff0150 end: 3fff03b0 offset: 01a0

>>>stack>>>
3fff02f0:  3ffeee20 00000000 3ffef2c4 401003e9  
3fff0300:  3ffef2a0 000000dd 000000dd 4010020c  
3fff0310:  3ffef2c4 3ffeee20 3fff12bc 401008e0  
3fff0320:  3ffe0111 3ffef2c4 3fff0364 40205f84  
3fff0330:  3ffeee20 3ffef2c4 00000128 402046c2  
3fff0340:  3ffef2c4 3ffeee20 3ffeee70 40202cdf  
3fff0350:  fc030df2 00000001 00010001 00000000  
3fff0360:  00fd000d 3fff16d4 3ffeef50 402026fe  
3fff0370:  00000009 feefeffe 40206258 40206240  
3fff0380:  00000000 00000000 00000001 3ffef380  
3fff0390:  3fffdad0 00000000 3ffef378 40206078  
3fff03a0:  feefeffe feefeffe 3ffef390 40100970  
<<<stack<<<
```

My current library code is commited to GitHub:
[https://github.com/mwildbolz/IRremoteESP8266/commit/52ee828d251ab73717b6df78d52f6d347b349370](url)

crankyoldgit commented:
Very interesting. Are you using dumpV2 to decode? Or some other code?

On Thu., 16 Nov. 2017, 4:46 am mwildbolz, <notifications@github.com> wrote:

> @crankyoldgit <https://github.com/crankyoldgit> I implemented this with a
> minor change (RPT Mark timing) and was able to decode a real remote signal.
> But there is an issue with repeated decoding.
> If i send a second signal with my remote, i get a memory alloc problem on
> my ESP causing it to restart, it just works once?!?
>
> check_poison_block is called for free block 0x3fff12b8
> Fatal exception 3(LoadStoreErrorCause):
> epc1=0x4010011d, epc2=0x00000000, epc3=0x00000000, excvaddr=0x40031af8, depc=0x00000000
>
> Exception (3):
> epc1=0x4010011d epc2=0x00000000 epc3=0x00000000 excvaddr=0x40031af8 depc=0x00000000
>
> ctx: cont
> sp: 3fff0150 end: 3fff03b0 offset: 01a0
>
> >>>stack>>>
> 3fff02f0:  3ffeee20 00000000 3ffef2c4 401003e9
> 3fff0300:  3ffef2a0 000000dd 000000dd 4010020c
> 3fff0310:  3ffef2c4 3ffeee20 3fff12bc 401008e0
> 3fff0320:  3ffe0111 3ffef2c4 3fff0364 40205f84
> 3fff0330:  3ffeee20 3ffef2c4 00000128 402046c2
> 3fff0340:  3ffef2c4 3ffeee20 3ffeee70 40202cdf
> 3fff0350:  fc030df2 00000001 00010001 00000000
> 3fff0360:  00fd000d 3fff16d4 3ffeef50 402026fe
> 3fff0370:  00000009 feefeffe 40206258 40206240
> 3fff0380:  00000000 00000000 00000001 3ffef380
> 3fff0390:  3fffdad0 00000000 3ffef378 40206078
> 3fff03a0:  feefeffe feefeffe 3ffef390 40100970
> <<<stack<<<
>
> My current library code is commited to GitHub:
>
> https://github.com/mwildbolz/IRremoteESP8266/commit/52ee828d251ab73717b6df78d52f6d347b349370
> <http://url>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/markszabo/IRremoteESP8266/pull/340#issuecomment-344689429>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMInDDp94iM5V5LFDuvYvAXQKjvDLwLAks5s2zGFgaJpZM4Qetr0>
> .
>

crankyoldgit committed:
Update Toshiba timings based on user feedback.
* @mwildbolz reported different timings were needed to capture correctly.
* fix a typo.

crankyoldgit commented:
Thanks @mwildbolz for the feedback. I've altered the timings to something I think is better. Can you please verify it works?
Can you please try this using this branch's [IRrecvDumpV2.ino](https://github.com/markszabo/IRremoteESP8266/blob/master/examples/IRrecvDumpV2/IRrecvDumpV2.ino) and see if that works (detects it) and doesn't crash on subsequent presses etc?

Also, could you please give me a copy of the raw data from one of your successful captures? That will be useful to add as a unit test case, and potentially allow me to reproduce the problems/issues you are encountering.

As for the crash, given your own repository reference, it looks like you've modified the code and are using something else (non-standard) to call the code. It may be the rest of the code or your changes which may be causing the crash. I've left a note in your branch to where I think the problem is, but without the rest of the code, I can't be sure what is causing your problem, nor can I reproduce it without it.

crankyoldgit committed:
Add validation of the checksum for decoding Toshiba A/C messages.
* Add a calcChecksum() method that can be called from outside the class.
* Add optional lengths to the checksum routines, just in case.
* Unit tests for the above changes.

crankyoldgit committed:
BugFix: useRaw mode of getMode() not working correctly.
I had a brain fade and used the wrong bit mask.
Added unit tests to cover this code & issue.

crankyoldgit committed:
Redo how the checksum verification is done to make the code more readable.

crankyoldgit committed:
Toshiba A/C always has at least one repeat.
According to @mwildbolz Toshiba A/C messages are sent at least twice.
Updated accordingly.

crankyoldgit committed:
Add outputing the state to a human readable format.

markszabo commented on src/ir_Toshiba.cpp:
```diff
@@ -25,8 +28,7 @@
 #define TOSHIBA_AC_ZERO_SPACE   472U
-#define TOSHIBA_AC_RPT_MARK     440U
-#define TOSHIBA_AC_RPT_SPACE   7048U
+#define TOSHIBA_AC_MIN_GAP   7048U
```
You might want to add an additional space here, just to keep the indentation.

markszabo reviewed (approved):
LGTM, one small comment :)

crankyoldgit committed:
Typo: Fix an indent issue.

crankyoldgit commented:
@markszabo Fixed. Curse you, _search & replace_! ;-)

crankyoldgit commented:
@mwildbolz I'm going to merge this PR into the v2.3.0-dev branch, but don't let that stop you from comment on this PR or logging a new issue etc. I still want your feedback and comments etc. I've just got some other code I want to get merged that depends on some of the changes here.

mwildbolz commented:
I tested the code with IRrecvDumpV2 and it worked :-)
The problem from yesterday was caused by a loose connection on hardware side - interesting, that it always worked once but not the second time (I tested 10 times).

Attached some real time data for the test cases
[testData.txt](https://github.com/markszabo/IRremoteESP8266/files/1479617/testData.txt)
